packages/jsii-diff/lib/diagnostics.ts

```diff
@@ -36,13 +36,48 @@ export function hasErrors(diags: Diagnostic[]) {
   return diags.some((diag) => diag.level === DiagLevel.Error);
 }
 
+export function onlyErrors(diags: Diagnostic[]) {
+  return diags.filter((diag) => diag.level === DiagLevel.Error);
+}
+
+export function onlyWarnings(diags: Diagnostic[]) {
+  return diags.filter((diag) => diag.level === DiagLevel.Warning);
+}
+
+export interface StabilityClassifications {
+  [Stability.Stable]?: boolean;
+  [Stability.Experimental]?: boolean;
+  [Stability.External]?: boolean;
+  [Stability.Deprecated]?: boolean;
+}
+
+export function errorsForStability(
+  errors: string[],
+  deprecatedExperimentalErrors = false,
+): StabilityClassifications {
+  return {
+    [Stability.Stable]: shouldError(errors, 'stable'),
+    [Stability.Experimental]:
+      shouldError(errors, 'experimental') || deprecatedExperimentalErrors,
+    [Stability.External]: shouldError(errors, 'external'),
+    [Stability.Deprecated]: shouldError(errors, 'deprecated'),
```

Comment:
I still feel that deprecated should also error when the class is 'stable'. That's usually what you want; if you don't, there are no guarantees on an API marked 'deprecated' and that's not correct, they still need to be preserved.

In my head we'll have some names for "error classes" as well which are predefined combinations of the standard stabilities. Also in my head we'd overload the names (so `error class STABLE = stabilities [STABLE, DEPRECATED]`), but that will most likely get confusing and not make sense to anyone but me :D. How about we introduce new names for the error classes, and have error classes for each of the stabilities as well that map to the stability itself. I think it looks like this:

![image](https://user-images.githubusercontent.com/524162/235601537-f619f26c-2dce-45f6-ab7e-1b01e61e5b32.png)

-------------

I also think the mapping will probably also be easier to understand if we make it explicit in the code:

```ts
type ErrorClass = 'all' | 'default' | 'stable' | 'experimental' | 'nonexperimental' | ...;

const ERRORCLASSES_TO_STABILITIES: Record<ErrorClass, Stability[]> = {
  'all': [Stability.Stable, Stability.Experimental, Stability.External, Stability.Deprecated],
  'default': [Stability.Stable, Stability.Deprecated],
  'nonexperimental': [Stability.Stable, Stability.Deprecated, Stability.External],
  'stable': [Stability.Stable],
  'experimental': [Stability.Experimental],
};
```

This is a bit more repetition, but extremely easy to parse and add new error classes to later.

Reply:
> I also think the mapping will probably also be easier to understand if we make it explicit in the code:

Yes, this much nicer. Thanks!

Reply:
jsii strictly enforces `deprecated` as its own stability.

For example this code:
```ts
    /**
     * @stability stable
     * @deprecated whatever
     **/
    export class FooBar { }
```

Will cause jsii to fail with: 
```console
      index.ts:6:5 - error JSII7999: @deprecated tag requires '@stability deprecated' or no @stability at all.
      
      6     export class FooBar { }
            ~~~~~~~~~~~~~~~~~~~~~~~
```

While this might be unintuitive to users, at least it's consistently enforced.
You can't have a `deprecated` feature that's also `stable` (or `experimental` or `external` for that matter).
Therefore defining `deprecated ∈ stable` would not be correct.

--- 

The current implementation doesn't allow anyone to make a decision about `deprecated` or `external` stabilities.
It looks more like this:

This is how the API currently treats stabilities:
![sketch1683022214169](https://user-images.githubusercontent.com/379814/235641593-4c393d8e-3e41-4713-a18c-6ee741c6f4d1.png)
With `external` and `deprecated` only every being a `WARNING`

Now in my mind the onion isn't quite as neatly defined as yours.
To me the whole thing looks more like this:
![sketch1683022165565](https://user-images.githubusercontent.com/379814/235642341-ac85f67f-6e03-441d-a047-f059c1aae77f.png)

---

In summary I think we agree on having pre-defined groups of stability classes. We probably agree on `all` as well.
I think you might also be okay with mix-and-matching groups together (`--error-on=stable --error-on=external`).

What we likely don't agree on yet is how to treat `deprecated` and any other pre-defined groups.
I don't have strong feelings around these additional groups, as long as the names are unambiguous (which might be tricky).

What I do disagree with is the proposed `default` group. Firstly, because the name is descriptive. It should have a descriptive name. Secondly because making it the implied default would constitute a breaking change. It do think it would actually make a reasonable default if we were to break `jsii-diff` out into it's own repo and release a v2.